## TRAVAIL SUR EAU DU ROBINET 

In [2]:
import pandas as pd
pd.set_option("display.max_columns", None)  # show all cols
pd.set_option("display.max_colwidth", None)  # show full width of showing cols
pd.set_option(
    "display.expand_frame_repr", False
)

In [6]:
edr_udi_res = pd.read_csv(r'C:\Users\EPPINGER Muriel\Documents\D4G_13\13_pollution_eau\analytics\EDR\UDI_RES_202412.txt', delimiter=',')


C:\Users\EPPINGER Muriel\AppData\Local\Temp\ipykernel_9072\1325013356.py:1: DtypeWarning: Columns (0,1,2,3,5,8) have mixed types. Specify dtype option on import or set low_memory=False.
  edr_udi_res = pd.read_csv(r'C:\Users\EPPINGER Muriel\Documents\D4G_13\13_pollution_eau\analytics\EDR\UDI_RES_202412.txt', delimiter=',')


In [5]:
edr_udi_plv = pd.read_csv(r'C:\Users\EPPINGER Muriel\Documents\D4G_13\13_pollution_eau\analytics\EDR\UDI_PLV_202412.txt', delimiter=',')

C:\Users\EPPINGER Muriel\AppData\Local\Temp\ipykernel_9072\1076701354.py:1: DtypeWarning: Columns (0,3,4,6,11,13,23,26) have mixed types. Specify dtype option on import or set low_memory=False.
  edr_udi_plv = pd.read_csv(r'C:\Users\EPPINGER Muriel\Documents\D4G_13\13_pollution_eau\analytics\EDR\UDI_PLV_202412.txt', delimiter=',')


In [5]:
# période de la BDD EDR_UDI

print(edr_udi_plv['dateprel'].min())
print(edr_udi_plv['dateprel'].max())

2016-04-05
2024-12-31


In [6]:
# extraction d'une referenceprel dnas bdd "prélèvements"

edr_udi_plv.loc[edr_udi_plv['referenceprel']=='00100141659',:]

distrcode     distrlib    distrsiret cddept inseecommune nomcommune   cdreseau              nomreseau codetypeinstallation    nomtypeinstallation inae    cdpointsurv           nompointsurv referenceprel    dateprel heureprel finaliteprel                                                                                 conclusionprel cdtypeeau                  libtypeeau plvconformitebacterio plvconformitechimique plvcomplet plvcdorganismepayeur typevisite plvcondition       preleveur   distrsiren plvconformiterefbacterio plvconformiterefchimique  cddeptsandre  dtautorisation  champcaptant  codebrgm  codebss     dmajplv   dmodifplv   dmodifins representativite     coord_x      coord_y
27  00101002  SOGEDO LYON  3.011928e+13    001        01032  BELIGNEUX  001000487  BELIGNEUX LA VALBONNE                  UDI  UNITE DE DISTRIBUTION  ESO  0010000000529  BELIGNEUX LA VALBONNE   00100141659  2024-08-12     11h56           CS  Eau d'alimentation conforme aux limites de qualité et non conforme aux références de qualité.        T   EAU DISTRIBUEE DESINFECTEE                     C                     C          O             00100973  D1                 NaN  13000807100107  301192803.0                        C                        N             1             NaN           NaN       NaN      NaN  2024-12-03  2024-12-04  2024-10-10                O  865227.563  6530365.073

In [7]:
# modification format reference prel pour comparaison avec autres bdd

edr_udi_res['referenceprel']=edr_udi_res['referenceprel'].astype('string')
edr_udi_res['referenceprel']=edr_udi_res['referenceprel'].apply(lambda x : '00'+x if len(x) == 9
                                                                 else '0'+x  if len(x) == 10
                                                                 else x)

In [8]:
# extraction d'un prélèvement dans bdd resultats

edr_udi_res.loc[edr_udi_res['referenceprel']=='00100141659',:]

cddept referenceprel    cdanalyse refechantillonlabo datereceptionechant cdlaboratoire   laboratoire cdtypeanalyse analysecdorganisme analysedate insituana  cdparametre cdparametresiseeaux  rqana rssigne   rsana cdunitereferencesiseeaux cdunitereference  cdfractionanalyse  sandrelaboratoire representativite
0       1   00100141659  100150585.0      LSE2408-35603          2024-08-12        6901.0  4.105453e+13        0011DN           100973.0  2024-11-27         L       6489.0                 ASP      0     NaN    0.00               SANS OBJET                X                NaN              217.0                O
1       1   00100141659  100150585.0      LSE2408-35603          2024-08-12        6901.0  4.105453e+13        0011DN           100973.0  2024-11-27         L       1042.0                BSIR     <1       <    0.00                n/(100mL)              226                NaN              217.0                O
2       1   00100141659  100150585.0      LSE2408-35603          2024-08-12        6901.0  4.105453e+13        0011DN           100973.0  2024-11-27         L       1303.0               CDT25    554     NaN  554.00                    µS/cm              147                NaN              217.0                O
3       1   00100141659          NaN                NaN                 NaN           NaN           NaN           NaN                NaN         NaN         T       1398.0              CL2LIB   0,12     NaN    0.12                mg(Cl2)/L              165                NaN                NaN                O
4       1   00100141659          NaN                NaN                 NaN           NaN           NaN           NaN                NaN         NaN         T       1399.0              CL2TOT   0,16     NaN    0.16                mg(Cl2)/L              165                NaN                NaN                O
5       1   00100141659          NaN                NaN                 NaN           NaN           NaN           NaN                NaN         NaN         T       1754.0                CLO2   N.M.       N     NaN                     mg/L              162                NaN                NaN                O
6       1   00100141659  100150585.0      LSE2408-35603          2024-08-12        6901.0  4.105453e+13        0011DN           100973.0  2024-11-27         L       1309.0               COULF     <5       <    0.00                 mg(Pt)/L              178                NaN              217.0                O
7       1   00100141659          NaN                NaN                 NaN           NaN           NaN           NaN                NaN         NaN         T       5900.0               COULQ      0     NaN    0.00               SANS OBJET                X                NaN                NaN                O
8       1   00100141659  100150585.0      LSE2408-35603          2024-08-12        6901.0  4.105453e+13        0011DN           100973.0  2024-11-27         L       1447.0                 CTF     <1       <    0.00                n/(100mL)              226                NaN              217.0                O
9       1   00100141659  100150585.0      LSE2408-35603          2024-08-12        6901.0  4.105453e+13        0011DN           100973.0  2024-11-27         L       1449.0               ECOLI     <1       <    0.00                n/(100mL)              226                NaN              217.0                O
10      1   00100141659  100150585.0      LSE2408-35603          2024-08-12        6901.0  4.105453e+13        0011DN           100973.0  2024-11-27         L       5440.0             GT22_68     <1       <    0.00                     n/mL              222                NaN              217.0                O
11      1   00100141659  100150585.0      LSE2408-35603          2024-08-12        6901.0  4.105453e+13        0011DN           100973.0  2024-11-27         L       5441.0             GT36_44     <1       <    0.00                     n/mL              222   

In [9]:
# recherche de doublons dans la base des résultats

doublons = edr_udi_res.duplicated()
df_doublons_res = edr_udi_res[doublons]
print(df_doublons_res)

Empty DataFrame
Columns: [cddept, referenceprel, cdanalyse, refechantillonlabo, datereceptionechant, cdlaboratoire, laboratoire, cdtypeanalyse, analysecdorganisme, analysedate, insituana, cdparametre, cdparametresiseeaux, rqana, rssigne, rsana, cdunitereferencesiseeaux, cdunitereference, cdfractionanalyse, sandrelaboratoire, representativite]
Index: []


In [10]:
# recherche de doublons dans la base des prélevements

doublons = edr_udi_plv.duplicated()
df_doublons_plv = edr_udi_plv[doublons]
print(df_doublons_plv)

Empty DataFrame
Columns: [distrcode, distrlib, distrsiret, cddept, inseecommune, nomcommune, cdreseau, nomreseau, codetypeinstallation, nomtypeinstallation, inae, cdpointsurv, nompointsurv, referenceprel, dateprel, heureprel, finaliteprel, conclusionprel, cdtypeeau, libtypeeau, plvconformitebacterio, plvconformitechimique, plvcomplet, plvcdorganismepayeur, typevisite, plvcondition, preleveur, distrsiren, plvconformiterefbacterio, plvconformiterefchimique, cddeptsandre, dtautorisation, champcaptant, codebrgm, codebss, dmajplv, dmodifplv, dmodifins, representativite, coord_x, coord_y]
Index: []


#Cas particulier : 1 prélèvement avec 2 codes analyse différents

In [11]:
edr_udi_plv.loc[edr_udi_plv['referenceprel']=='00100143183',:]

distrcode                              distrlib  distrsiret cddept inseecommune  nomcommune   cdreseau                   nomreseau codetypeinstallation    nomtypeinstallation inae    cdpointsurv                    nompointsurv referenceprel    dateprel heureprel finaliteprel                                                                                           conclusionprel cdtypeeau                  libtypeeau plvconformitebacterio plvconformitechimique plvcomplet plvcdorganismepayeur typevisite plvcondition       preleveur  distrsiren plvconformiterefbacterio plvconformiterefchimique  cddeptsandre  dtautorisation  champcaptant  codebrgm  codebss     dmajplv   dmodifplv   dmodifins representativite  coord_x  coord_y
35  00101000  CC TERRE VALSERHONE - REGIE DES EAUX         NaN    001        01033  VALSERHONE  001001033  BELLEGARDE QUARTIERS HAUTS                  UDI  UNITE DE DISTRIBUTION  ESO  0010000002222  BELLEGARDE STATION D'EPURATION   00100143183  2024-11-21     12h00           CS  Eau d'alimentation conforme aux exigences de qualité en vigueur pour l'ensemble des paramètres mesurés.        T   EAU DISTRIBUEE DESINFECTEE                     C                     C          O             00101000  D2                 NaN  13000807100107         NaN                        C                        C             1             NaN           NaN       NaN      NaN  2024-12-03  2024-12-04  2019-12-04                O      NaN      NaN

In [12]:
edr_udi_res.loc[((edr_udi_res['referenceprel']==100143183) & (edr_udi_res['cdparametresiseeaux']=="ASP")),:]

Empty DataFrame
Columns: [cddept, referenceprel, cdanalyse, refechantillonlabo, datereceptionechant, cdlaboratoire, laboratoire, cdtypeanalyse, analysecdorganisme, analysedate, insituana, cdparametre, cdparametresiseeaux, rqana, rssigne, rsana, cdunitereferencesiseeaux, cdunitereference, cdfractionanalyse, sandrelaboratoire, representativite]
Index: []

In [13]:
# création d'une liste unique de referenceprel

edr_plv = edr_udi_plv[['referenceprel']].copy()
unique_edr_plv=edr_plv.drop_duplicates()
unique_edr_plv.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24532 entries, 0 to 24531
Data columns (total 1 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   referenceprel  24532 non-null  object
dtypes: object(1)
memory usage: 191.8+ KB


In [14]:
# recherche liste unique referenceprel/cdparametres/SISEEAUX

ref_param = edr_udi_res[['referenceprel','cdparametre','cdparametresiseeaux']].copy()
unique_edr_param=ref_param.drop_duplicates()
unique_edr_param.info()

<class 'pandas.core.frame.DataFrame'>
Index: 578730 entries, 0 to 587306
Data columns (total 3 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   referenceprel        578730 non-null  object 
 1   cdparametre          577961 non-null  float64
 2   cdparametresiseeaux  578679 non-null  object 
dtypes: float64(1), object(2)
memory usage: 17.7+ MB


## TRAVAIL SUR EAU DE COMMUNE

In [15]:
import duckdb
from pipelines.tasks._common import DUCKDB_FILE

con = duckdb.connect(database=DUCKDB_FILE, read_only=True)

In [16]:
query_2024 = """
SELECT * from edc_prelevements
"""

edc_plv = con.sql(query_2024)
data_edc_plv = edc_plv.df()
data_edc_plv.head(2)

cddept   cdreseau inseecommuneprinc nomcommuneprinc cdreseauamont nomreseauamont pourcentdebit referenceprel   dateprel heureprel                                                                                           conclusionprel                         ugelib                        distrlib                          moalib plvconformitebacterio plvconformitechimique plvconformitereferencebact plvconformitereferencechim  de_partition de_ingestion_date
0    001  001000003             01007        AMBRONAY          None           None          None   00100087319 2016-02-22     11h07  Eau d'alimentation conforme aux exigences de qualité en vigueur pour l'ensemble des paramètres mesurés.  SI REGION D'AMBERIEU-EN-BUGEY  SIE REGION D'AMBERIEU-EN-BUGEY  SIE REGION D'AMBERIEU-EN-BUGEY                     C                     C                          C                          C          2016        2025-02-08
1    001  001000003             01007        AMBRONAY          None           None          None   00100087349 2016-02-22     10h21  Eau d'alimentation conforme aux exigences de qualité en vigueur pour l'ensemble des paramètres mesurés.  SI REGION D'AMBERIEU-EN-BUGEY  SIE REGION D'AMBERIEU-EN-BUGEY  SIE REGION D'AMBERIEU-EN-BUGEY                     C                     C                          C                          C          2016        2025-02-08

In [17]:
exple = """
SELECT * from edc_prelevements
WHERE referenceprel='00100143183'
"""

edc_plv_exple = con.sql(exple)
data_edc_plv_exple = edc_plv_exple.df()
data_edc_plv_exple

cddept   cdreseau inseecommuneprinc nomcommuneprinc cdreseauamont nomreseauamont pourcentdebit referenceprel   dateprel heureprel                                                                                           conclusionprel                       ugelib                              distrlib                                moalib plvconformitebacterio plvconformitechimique plvconformitereferencebact plvconformitereferencechim  de_partition de_ingestion_date
0    001  001001033             01033      VALSERHONE          None           None          None   00100143183 2024-11-21     12h00  Eau d'alimentation conforme aux exigences de qualité en vigueur pour l'ensemble des paramètres mesurés.  CC PAYS BELLEGARDIEN (CCPB)  CC TERRE VALSERHONE - REGIE DES EAUX  CC TERRE VALSERHONE - REGIE DES EAUX                     C                     C                          C                          C          2024        2025-02-08

La referenceprel est identique entre EDR et EDC et retourne les mêmes données dans la base "prélèvements"

In [18]:
exple_res = """
SELECT * from edc_resultats
WHERE referenceprel='00100143183'
"""

edc_res_exple = con.sql(exple_res)
data_edc_res_exple = edc_res_exple.df()
data_edc_res_exple

cddept referenceprel cdparametresiseeaux  cdparametre                                         libmajparametre                                         libminparametre libwebparametre qualitparam insituana                     rqana cdunitereferencesiseeaux cdunitereference     limitequal                refqual  valtraduite    casparam referenceanl  de_partition de_ingestion_date
0     001   00100143183                ACPT       1453.0                                             ACÉNAPHTÈNE                                             Acénaphtène            None           N         L                    <0,001                     µg/L              133           None                   None        0.000     83-32-9  00100152188          2024        2025-02-08
1     001   00100143183               ACRYL       1457.0                                              ACRYLAMIDE                                              Acrylamide            None           N         L                     <0,10                     µg/L              133     <=0.1 µg/L                   None        0.000     79-06-1  00100152188          2024        2025-02-08
2     001   00100143183              ANTHRA       1458.0                                              ANTHRACÈNE                                              Anthracène            None           N         L                    <0,001                     µg/L              133           None                   None        0.000    120-12-7  00100152188          2024        2025-02-08
3     001   00100143183                 ASP       6489.0                                     ASPECT (QUALITATIF)                                     Aspect (qualitatif)            None           O         L             Aspect normal               SANS OBJET                X           None                   None        0.000        None  00100152188          2024        2025-02-08
4     001   00100143183                 ASP       6489.0                                     ASPECT (QUALITATIF)                                     Aspect (qualitatif)            None           O         L             Aspect normal               SANS OBJET                X           None                   None        0.000        None  00100152189          2024        2025-02-08
5     001   00100143183               BAPYR       1115.0                                        BENZO(A)PYRÈNE *                                        Benzo(a)pyrène *            None           N         L                   <0,0001                     µg/L              133    <=0.01 µg/L                   None        0.000     50-32-8  00100152188          2024        2025-02-08
6     001   00100143183              BBFLUO       1116.0                                    BENZO(B)FLUORANTHÈNE                                    Benzo(b)fluoranthène            None           N         L                   <0,0005                     µg/L              133     <=0.1 µg/L                   None        0.000    205-99-2  00100152188          2024        2025-02-08
7     001   00100143183              BENZAN       1082.0                                          BENZANTHRACÈNE                                          Benzanthracène            None           N         L                    <0,001                     µg/L              133           None                   None        0.000     56-55-3  00100152188          2024        2025-02-08
8     001   00100143183              BGPERY       1118.0                                    BENZO(G,H,I)PÉRYLÈNE                                    Benzo(g,h,i)pérylène            None           N         L                  <0,00050                     µg/L              133     <=0.1 µg/L                   None        0.000    191-24-2  00100152188          2024        2025-02-08
9     001   00100143183              BKFLUO       1117.0                                    BENZO(K)FLUORANTHÈNE                                    Benzo(k)fluoranthène     

In [19]:
# recherche de doublons dans la base des prélevements 

doublons = data_edc_plv.duplicated()
edc_doublons_plv = data_edc_plv[doublons]
print(edc_doublons_plv)

Empty DataFrame
Columns: [cddept, cdreseau, inseecommuneprinc, nomcommuneprinc, cdreseauamont, nomreseauamont, pourcentdebit, referenceprel, dateprel, heureprel, conclusionprel, ugelib, distrlib, moalib, plvconformitebacterio, plvconformitechimique, plvconformitereferencebact, plvconformitereferencechim, de_partition, de_ingestion_date]
Index: []


In [20]:
# recherche liste unique referenceprel

edc_plv = data_edc_plv[['referenceprel']].copy()
unique_edc_plv=edr_plv.drop_duplicates()
unique_edc_plv.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24532 entries, 0 to 24531
Data columns (total 1 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   referenceprel  24532 non-null  object
dtypes: object(1)
memory usage: 191.8+ KB


Le nb de prélèvements est identique entre EDC et EDR dans les base prélèvements

## COMPARAISON EDC ET EDC

In [21]:
# recherche des prélèvements présents dans EDC et absents dans EDR

resultat = unique_edc_plv[~unique_edc_plv['referenceprel'].isin(unique_edr_plv['referenceprel'])]
print(resultat)

Empty DataFrame
Columns: [referenceprel]
Index: []


In [22]:
# recherche des prélèvements présents dans EDR et absents dans EDC

resultat2 = unique_edr_plv[~unique_edr_plv['referenceprel'].isin(unique_edc_plv['referenceprel'])]
print(resultat2)

Empty DataFrame
Columns: [referenceprel]
Index: []


Les datasets EDC et EDR_UDI comportent les mêmes prélèvements
Vérifions maintenant s'ils comprennent les mêmes paramètres

In [23]:
# recherche nb prélèvements par paramètres EDR

unique_edr_param.groupby('cdparametresiseeaux')['referenceprel'].count()

cdparametresiseeaux
1112TCE     31
111TCL      90
1122TCE     23
112TCE      53
11DCET      53
          ... 
XYLMP        2
XYLO        17
ZCYINE      12
ZN          45
ZOXAMID    217
Name: referenceprel, Length: 1011, dtype: int64

In [32]:
unique_edr_param.describe(include='all')

referenceprel    cdparametre cdparametresiseeaux
count         578730  577961.000000              578679
unique         24531            NaN                1011
top      05100139536            NaN                  PH
freq             588            NaN               23092
mean             NaN    3053.822481                 NaN
std              NaN    3983.491044                 NaN
min              NaN    1034.000000                 NaN
25%              NaN    1309.000000                 NaN
50%              NaN    1447.000000                 NaN
75%              NaN    5441.000000                 NaN
max              NaN   99999.000000                 NaN

In [28]:
# recherche nb prélèvements par paramètres EDC

param_res = """
SELECT DISTINCT "referenceprel","cdparametre","cdparametresiseeaux" 
from edc_resultats

"""

edc_param_res = con.sql(param_res)
df_edc_param_res = edc_param_res.df()


FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

In [34]:
df_edc_param_res= df_edc_param_res.drop_duplicates()

In [31]:
df_edc_param_res.groupby('cdparametresiseeaux')['referenceprel'].count()

cdparametresiseeaux
004THIONA        1
006CFLERT        1
014CA-CAL      336
014CO2AG       211
021ANION       196
             ...  
ZIRA          7151
ZN            5794
ZOXAMID      88415
ZYGOPH1        324
ZYGOPHY        306
Name: referenceprel, Length: 1691, dtype: int64

In [35]:
df_edc_param_res.describe(include='all')

referenceprel   cdparametre cdparametresiseeaux
count      111322482  1.111687e+08           111322482
unique       2536393           NaN                1691
top      00400113259           NaN                TEAU
freq             760           NaN             2375216
mean             NaN  2.633498e+03                 NaN
std              NaN  2.809932e+03                 NaN
min              NaN  0.000000e+00                 NaN
25%              NaN  1.303000e+03                 NaN
50%              NaN  1.500000e+03                 NaN
75%              NaN  2.924000e+03                 NaN
max              NaN  9.999900e+04                 NaN

Dans base EDC : 1691 paramètres SISEEAUX contre 1011 dans base EDR

Les données de la base UDI EDR sont donc incluses dans base EDC

Reste à vérifer base CAP et TTP

Question : 
Dans EDC : il y a 24532 prélevement dans base prélèvements alors qu'il y aurait 2 536 393 referenceprel unique dans la base des résultats ?????